keps/sig-instrumentation/0034-distributed-tracing-kep.md

```diff
@@ -0,0 +1,179 @@
+
+
+### Non-Goals
+
+* To replace existing logging, metrics, or the events API
+* To trace operations from all Kubernetes resource types in a generic manner (i.e. without manual instrumentation)
+
+
+## Proposal
+
+### Architecture
+
+#### Trace lifecycle
+
+Kubernetes is unique in that it is constantly reconciling its actual state towards some desired state. As a result, it has no definitive concept of an "operation", which breaks the traditional model for distributed tracing. This raises the question of when to begin traces, and when to end them.
+
+In this proposal, we choose to _only_ trace phases of an object's lifecycle wherein it's correcting from an undesired state to its desired state, and to end the trace when it enters this desired state. This means that the same object will export traces for each reconciliation it undergoes. This decision was made because:
+
+1) These reconciliation phases are where Kubernetes performs actual work that can be traced
+2) Prevents traces from spanning the entire life of an object, which can be indefninite
+3) Provides a consistent, concrete duration for the root span which remains applicable across object types
+4) The time required for Kubernetes controllers to reconcile desired and actual state is, by definition, a measurement of either latency experienced by end-users in the fulfillment of their requests, or down-time due to a disruption
+
+#### Root spans
+
+In the standard model for distributed tracing, there exists a span in each trace that all other spans are descendents of and which extends the length of the entire trace, called the `root span`. 
+
+The Kubernetes component that kicks off an operation might not be the same component that ends it. In this proposal, when we are at the point where we want to end a root span, we craft a span to export which acts as the root span for the trace. For example, when the kubelet updates a pod from `Pending` to `Running`, it creates a root span using the start time of the pod as the start, and the current time as the end.
+
+
+#### Context propagation
+
+To correlate work done between components as belonging to the same trace, we must pass span context across process boundaries. In traditional distributed systems, this context can be passed down through RPC metadata or HTTP headers. Kubernetes, however, due to its watch-based nature, requires us to attach trace context directly to the target object. 
+
+In this proposal, we choose to propagate this span context as an encoded string an object annotation called `trace.kubernetes.io/context`. This annotation value is regenerated and replaced when an object's trace ends, to achieve the desired behavior from [section one](#trace-lifecycle). 
+
+This proposal chooses to use annotations as a less invasive alternative to adding a field to object metadata, but as this proposal matures, adding trace context to the official API should be considered. 
```

Comment:
I would like to see it mentioned very clearly that additing tracing will not result in increasing amount of requests to apiserver - otherwise it may visibly impact performance of the system, which we definitely don't want.

[That also implicitly means, that "end of an operation" has to be associated with some write request to apiserver, which I'm not 100% convinced will always be the case in cases that we're interested about].

@kubernetes/sig-scalability-api-reviews

Reply:
Thats a great point.  It definitely will have an impact during alpha, as we are using annotations. We can remove the extra write, at least in theory, if we moved context propagation in-tree by adding the ability to update/regenerate the trace context during a status update.

The "end of an operation" always coincides with a status update from a non-desired state to the desired state in the current proposal.  This implicitly means objects without a status don't receive new trace contexts outside of creation/update/deletion (i'm not convinced tracing is applicable to such objects).  Do you have a case in-mind you are not sure about?

Reply:
> It definitely will have an impact during alpha, as we are using annotations.

I think it depends. If e.g. we say that pod creation should start a span, then we can build that into the machinery so that it will be done automatically. So I actually don't agree it has to be the case.

And just to be clear on that: I can live with this requirement not being satisfied in Alpha state, but I'm not going to approve it for beta+ if it will generally be creating higher load on apiserver (there can be some exceptions for some rare flows, but in general it cannot cause additional writes).